packages/bson-bench/src/common.ts

```diff
@@ -0,0 +1,186 @@
+    }
+  }
+
+  /**
+   * returns the package if it exists, otherwise return undefined
+   */
+  check<B extends BSONLib>(): B | undefined {
+    try {
+      return require(this.computedModuleName);
+    } catch (e) {
+      return undefined;
+    }
+  }
+
+  /**
+   * Installs the package using the npm cli
+   *
+   * Note that this function should not be run in the same Node process that imports the installed
+   * module
+   **/
+  async install(): Promise<void> {
+    let source: string;
+    switch (this.type) {
+      case 'npm':
+        source = `npm:${this.library}@${this.npmVersion}`;
+        break;
+      case 'git':
+        switch (this.library) {
+          case 'bson':
+            source = `git://github.com/mongodb/js-bson#${this.gitCommitish}`;
+            break;
+          case 'bson-ext':
+            source = `git://github.com/mongodb-js/bson-ext#${this.gitCommitish}`;
+            break;
+        }
+        break;
+      case 'local':
+        source = `${this.localPath}`;
+        // Check if path exists as npm install will not throw an error if this is not the case
+        if (!(await exists(source))) throw new Error(`'${source}' not found`);
+        break;
+    }
+
+    const npmInstallProcess = cp.exec(
+      `npm install ${this.computedModuleName}@${source} --no-save`,
+      { encoding: 'utf8', cwd: __dirname }
+    );
+
+    const exitCode: number = (await once(npmInstallProcess, 'exit'))[0];
+    if (exitCode !== 0) {
+      throw new Error(`unable to install module: ${this.computedModuleName}@${source}`);
+    }
+  }
+}
+
+export interface Document {
+  [key: string]: any;
+}
+
+export interface BSONLib {
+  deserialize: (b: Uint8Array, options?: any) => Document;
+  serialize: (o: Document, options?: any) => Uint8Array;
+  EJSON?: any;
+}
+
+export type BenchmarkSpecification = {
+  /** Path to test document */
+  documentPath: string;
+  /** BSON operation to be benchmarked */
+  operation: 'serialize' | 'deserialize';
+  /** Options to be passed in to the operation being performed */
+  options: Record<string, any>;
+  /** Number of iterations that will be timed and used to calculate summary statistics*/
+  iterations: number;
+  /** Number of iterations that will be run to warm up the V8 engine */
+  warmup: number;
+  /** Specifier of the bson or bson-ext library to be used. Can be an npm package, git repository or
+   * local package */
+  library: string;
+};
+
+export interface IPCMessage {
+  pe: 'returnResult' | 'returnError' | 'runBenchmark';
+  benchmark?: BenchmarkSpecification;
+  result?: BenchmarkResult;
+  error?: Error;
+}
+
+export interface RunBenchmarkMessage extends IPCMessage {
+  type: 'runBenchmark';
+  benchmark: BenchmarkSpecification;
+}
+
+export interface ResultMessage extends IPCMessage {
+  type: 'returnResult';
+  result: BenchmarkResult;
+}
+
+export interface ErrorMessage extends IPCMessage {
+  type: 'returnError';
+  error: Error;
+}
```

Comment:
optional: Is it accurate that a ResultMessage could have a `benchmark` property?  Or an ErrorMessage could have a result?

Interface `extend`ing can lead to ambiguity like this.  An alternate design is to use a discriminated union - https://basarat.gitbook.io/typescript/type-system/discriminated-unions.  In this case, it would look something like this:

```typescript
export interface RunBenchmarkMessage {
  type: 'runBenchmark';
  benchmark: BenchmarkSpecification;
}

export interface ResultMessage {
  type: 'returnResult';
  result: BenchmarkResult;
}

export interface ErrorMessage {
  type: 'returnError';
  error: Error;
}

export type IPCMessage = RunBenchmarkMessage | ResultMessage | ErrorMessage;
```

Reply:
Good catch! That does clear up the ambiguity there.